File: apps/api/app/api/v1/endpoints/ai.py
"""
AI Copilot Proxy Endpoints — Phase 12.

Proxies requests to the AI Service, enforcing:
1. User authentication (JWT)
2. Workspace access verification
3. Interview participant check (only interviewers/hosts can access AI copilot, NOT candidates)
4. Telemetry audit retrieval
"""

import logging
import uuid
from typing import Any, Dict, List, Optional
from datetime import timedelta

from fastapi import APIRouter, Depends, HTTPException, Query, status
import httpx
from pydantic import BaseModel, Field
from sqlalchemy import desc, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import (
    get_current_user,
    get_db,
    verify_interview_access,
    verify_workspace_access,
)
from app.core.config import settings
from app.core.security import create_access_token
from app.models.ai import AIRequestLog
from app.models.interview import Interview, InterviewParticipant, ParticipantRole
from app.models.user import User, UserRole

logger = logging.getLogger("interviewos.api.ai")
router = APIRouter(prefix="/ai", tags=["AI Copilot & Gateway"])


# ---------------------------------------------------------------------------
# Request Schemas
# ---------------------------------------------------------------------------

class AITaskProxyRequest(BaseModel):
    interview_id: str
    workspace_id: str
    task_type: str
    session_id: Optional[str] = None
    input_data: Dict[str, Any] = Field(default_factory=dict)


class QuestionGenProxyRequest(BaseModel):
    interview_id: str
    workspace_id: str
    session_id: Optional[str] = None
    difficulty: str = "medium"
    question_type: str = "technical"
    topic_focus: Optional[str] = None
    topics_covered: List[str] = Field(default_factory=list)
    candidate: Optional[Dict[str, Any]] = None
    job: Optional[Dict[str, Any]] = None
    previous_questions: List[str] = Field(default_factory=list)
    current_stage: Optional[str] = None


class FollowUpProxyRequest(BaseModel):
    interview_id: str
    workspace_id: str
    question_asked: str
    candidate_answer: str
    session_id: Optional[str] = None
    target_skill: Optional[str] = None
    difficulty: str = "medium"
    remaining_seconds: int = 600


class ResumeAnalysisProxyRequest(BaseModel):
    interview_id: str
    workspace_id: str
    resume_text: Optional[str] = None
    candidate_profile: Optional[Dict[str, Any]] = None
    job: Optional[Dict[str, Any]] = None


class JobAnalysisProxyRequest(BaseModel):
    interview_id: str
    workspace_id: str
    job: Dict[str, Any]


class CodingAnalysisProxyRequest(BaseModel):
    interview_id: str
    workspace_id: str
    session_id: Optional[str] = None
    problem: Optional[Dict[str, Any]] = None
    candidate_code: Optional[str] = None
    execution_result: Optional[Dict[str, Any]] = None
    submission_history: List[Dict[str, Any]] = Field(default_factory=list)


class SystemDesignProxyRequest(BaseModel):
    interview_id: str
    workspace_id: str
    session_id: Optional[str] = None
    whiteboard_state: Optional[Dict[str, Any]] = None
    candidate_explanation: Optional[str] = None
    problem_statement: Optional[str] = None


# ---------------------------------------------------------------------------
# Helper: Verify Interviewer Access and Proxy to AI Service
# ---------------------------------------------------------------------------

async def _ensure_interviewer_access(
    interview_id_str: str,
    workspace_id_str: str,
    current_user: User,
    db: AsyncSession,
) -> None:
    """
    Ensure the caller has access to the workspace and is an interviewer.
    Candidates MUST be blocked with 403.
    """
    if current_user.role == UserRole.PLATFORM_ADMIN:
        return

    try:
        w_uuid = uuid.UUID(workspace_id_str)
        i_uuid = uuid.UUID(interview_id_str)
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid UUID format for workspace_id or interview_id",
        )

    # 1. Verify workspace membership
    await verify_workspace_access(w_uuid, current_user, db)

    # 2. Check if user is a candidate participant in this interview
    stmt = select(InterviewParticipant).where(
        InterviewParticipant.interview_id == i_uuid,
        InterviewParticipant.user_id == current_user.id,
    )
    res = await db.execute(stmt)
    participant = res.scalar_one_or_none()

    if participant and participant.role == ParticipantRole.CANDIDATE:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="AI Copilot features are restricted to interviewers",
        )


async def _proxy_ai_request(endpoint: str, payload: dict, current_user: User) -> dict:
    """Forward request to AI Service with minted JWT claims."""
    url = f"{settings.AI_SERVICE_URL.rstrip('/')}/{endpoint.lstrip('/')}"
    token = create_access_token(
        subject=str(current_user.id),
        extra_claims={
            "workspace_id": payload.get("workspace_id"),
            "role": current_user.role.value if hasattr(current_user.role, "value") else str(current_user.role),
            "is_interviewer": True,
        },
        expires_delta=timedelta(minutes=5),
    )

    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json",
    }

    try:
        async with httpx.AsyncClient(timeout=float(settings.AI_TIMEOUT_SECONDS)) as client:
            resp = await client.post(url, json=payload, headers=headers)
            if resp.status_code == status.HTTP_403_FORBIDDEN:
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail=resp.json().get("detail", "Access forbidden by AI service"),
                )
            if resp.status_code >= 400:
                logger.warning("AI service error %d: %s", resp.status_code, resp.text)
                return {
                    "status": "failed",
                    "error_message": f"AI service error ({resp.status_code})",
                    "output": None,
                }
            return resp.json()
    except httpx.RequestError as exc:
        logger.error("AI service unreachable at %s: %s", url, exc)
        return {
            "status": "failed",
            "error_message": f"AI service unreachable: {str(exc)}",
            "output": None,
        }


# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------

@router.post("/tasks")
async def run_ai_task(
    req: AITaskProxyRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await _ensure_interviewer_access(req.interview_id, req.workspace_id, current_user, db)
    return await _proxy_ai_request("ai/tasks", req.model_dump(), current_user)


@router.post("/generate-question")
async def generate_question(
    req: QuestionGenProxyRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await _ensure_interviewer_access(req.interview_id, req.workspace_id, current_user, db)
    return await _proxy_ai_request("ai/generate-question", req.model_dump(), current_user)


@router.post("/generate-follow-up")
async def generate_follow_up(
    req: FollowUpProxyRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await _ensure_interviewer_access(req.interview_id, req.workspace_id, current_user, db)
    return await _proxy_ai_request("ai/generate-follow-up", req.model_dump(), current_user)


@router.post("/analyze-resume")
async def analyze_resume(
    req: ResumeAnalysisProxyRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await _ensure_interviewer_access(req.interview_id, req.workspace_id, current_user, db)
    return await _proxy_ai_request("ai/analyze-resume", req.model_dump(), current_user)


@router.post("/analyze-job")
async def analyze_job(
    req: JobAnalysisProxyRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await _ensure_interviewer_access(req.interview_id, req.workspace_id, current_user, db)
    return await _proxy_ai_request("ai/analyze-job", req.model_dump(), current_user)


@router.post("/analyze-coding")
async def analyze_coding(
    req: CodingAnalysisProxyRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await _ensure_interviewer_access(req.interview_id, req.workspace_id, current_user, db)
    return await _proxy_ai_request("ai/analyze-coding", req.model_dump(), current_user)


@router.post("/analyze-system-design")
async def analyze_system_design(
    req: SystemDesignProxyRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await _ensure_interviewer_access(req.interview_id, req.workspace_id, current_user, db)
    return await _proxy_ai_request("ai/analyze-system-design", req.model_dump(), current_user)


@router.get("/health")
async def ai_health(current_user: User = Depends(get_current_user)):
    """Check AI service connectivity and provider status."""
    url = f"{settings.AI_SERVICE_URL.rstrip('/')}/health"
    try:
        async with httpx.AsyncClient(timeout=5.0) as client:
            resp = await client.get(url)
            return resp.json()
    except Exception as exc:
        return {
            "status": "degraded",
            "service": "ai-service",
            "error": str(exc),
        }


@router.get("/logs")
async def get_ai_logs(
    workspace_id: str = Query(...),
    interview_id: Optional[str] = Query(None),
    limit: int = Query(50, ge=1, le=200),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Retrieve telemetry log for AI requests within a workspace."""
    try:
        w_uuid = uuid.UUID(workspace_id)
    except ValueError:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid workspace UUID")

    await verify_workspace_access(w_uuid, current_user, db)

    query = (
        select(AIRequestLog)
        .where(AIRequestLog.workspace_id == workspace_id)
        .order_by(desc(AIRequestLog.created_at))
        .limit(limit)
    )
    if interview_id:
        query = query.where(AIRequestLog.interview_id == interview_id)

    res = await db.execute(query)
    logs = res.scalars().all()

    return [
        {
            "id": str(log.id),
            "request_id": log.request_id,
            "workspace_id": log.workspace_id,
            "interview_id": log.interview_id,
            "agent_name": log.agent_name,
            "task_type": log.task_type,
            "provider": log.provider,
            "model": log.model,
            "is_fallback": log.is_fallback,
            "input_tokens": log.input_tokens,
            "output_tokens": log.output_tokens,
            "total_tokens": log.total_tokens,
            "latency_ms": log.latency_ms,
            "success": log.success,
            "error_type": log.error_type,
            "created_at": log.created_at.isoformat() if log.created_at else None,
        }
        for log in logs
    ]

File: apps/web/lib/api/ai.ts
import { apiClient } from '../api';

export interface GeneratedQuestionResponse {
  request_id: string;
  task_type: string;
  output?: {
    question_text: string;
    question_type: string;
    difficulty: string;
    target_skills: string[];
    rubric: Array<{
      criteria: string;
      level_4_exemplary: string;
      level_3_competent: string;
      level_2_developing: string;
      level_1_unsatisfactory: string;
    }>;
    follow_up_hooks: string[];
    suggested_time_minutes: number;
    expected_key_points: string[];
  };
  metadata?: {
    model: string;
    provider: string;
    tokens: number;
    latency_ms: number;
    is_fallback: boolean;
  };
  status?: string;
  error_message?: string;
}

export interface FollowUpResponse {
  request_id: string;
  task_type: string;
  output?: {
    primary_follow_up: string;
    alternative_angles: string[];
    difficulty: string;
    skill_targeted: string;
    purpose: string;
    time_budget_minutes: number;
  };
  metadata?: {
    model: string;
    provider: string;
    tokens: number;
    latency_ms: number;
    is_fallback: boolean;
  };
  status?: string;
  error_message?: string;
}

export interface ResumeAnalysisResponse {
  request_id: string;
  task_type: string;
  output?: {
    skills_claimed: string[];
    key_experience_themes: string[];
    depth_verification_questions: Array<{
      topic: string;
      question: string;
      look_for: string;
    }>;
    project_deep_dives: Array<{
      project_name: string;
      key_challenges_to_probe: string[];
      suggested_scenario: string;
    }>;
    skill_gaps_or_red_flags: string[];
    seniority_assessment: string;
  };
  metadata?: {
    model: string;
    provider: string;
    tokens: number;
    latency_ms: number;
    is_fallback: boolean;
  };
  status?: string;
  error_message?: string;
}

export interface CodingAnalysisResponse {
  request_id: string;
  task_type: string;
  output?: {
    approach_classification: string;
    time_complexity: string;
    space_complexity: string;
    edge_cases_unhandled: string[];
    code_smells: string[];
    suggested_hints: string[];
    follow_up_optimizations: string[];
    sandbox_result_summary: string;
  };
  metadata?: {
    model: string;
    provider: string;
    tokens: number;
    latency_ms: number;
    is_fallback: boolean;
  };
  status?: string;
  error_message?: string;
}

export interface SystemDesignAnalysisResponse {
  request_id: string;
  task_type: string;
  output?: {
    components_identified: string[];
    architecture_style: string;
    single_points_of_failure: string[];
    scalability_concerns: string[];
    missing_elements: string[];
    targeted_critique_questions: string[];
    next_drill_down: string;
  };
  metadata?: {
    model: string;
    provider: string;
    tokens: number;
    latency_ms: number;
    is_fallback: boolean;
  };
  status?: string;
  error_message?: string;
}

export interface AILogEntry {
  id: string;
  request_id: string;
  workspace_id: string;
  interview_id?: string;
  agent_name?: string;
  task_type?: string;
  provider: string;
  model: string;
  is_fallback: boolean;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  latency_ms: number;
  success: boolean;
  error_type?: string;
  created_at: string;
}

export const aiApi = {
  generateQuestion: (data: {
    interview_id: string;
    workspace_id: string;
    session_id?: string;
    difficulty?: string;
    question_type?: string;
    topic_focus?: string;
    topics_covered?: string[];
    candidate?: any;
    job?: any;
    previous_questions?: string[];
    current_stage?: string;
  }) => {
    return apiClient<GeneratedQuestionResponse>('/ai/generate-question', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  generateFollowUp: (data: {
    interview_id: string;
    workspace_id: string;
    question_asked: string;
    candidate_answer: string;
    session_id?: string;
    target_skill?: string;
    difficulty?: string;
    remaining_seconds?: number;
  }) => {
    return apiClient<FollowUpResponse>('/ai/generate-follow-up', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  analyzeResume: (data: {
    interview_id: string;
    workspace_id: string;
    resume_text?: string;
    candidate_profile?: any;
    job?: any;
  }) => {
    return apiClient<ResumeAnalysisResponse>('/ai/analyze-resume', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  analyzeCoding: (data: {
    interview_id: string;
    workspace_id: string;
    session_id?: string;
    problem?: any;
    candidate_code?: string;
    execution_result?: any;
    submission_history?: any[];
  }) => {
    return apiClient<CodingAnalysisResponse>('/ai/analyze-coding', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  analyzeSystemDesign: (data: {
    interview_id: string;
    workspace_id: string;
    session_id?: string;
    whiteboard_state?: any;
    candidate_explanation?: string;
    problem_statement?: string;
  }) => {
    return apiClient<SystemDesignAnalysisResponse>('/ai/analyze-system-design', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  getHealth: () => {
    return apiClient<{ status: string; service: string; providers?: any }>('/ai/health', {
      method: 'GET',
    });
  },

  getLogs: (workspaceId: string, interviewId?: string, limit: number = 50) => {
    const params: Record<string, string> = {
      workspace_id: workspaceId,
      limit: limit.toString(),
    };
    if (interviewId) {
      params.interview_id = interviewId;
    }
    return apiClient<AILogEntry[]>('/ai/logs', {
      method: 'GET',
      params,
    });
  },

  // Phase 14 Adaptive Endpoints
  ingestTranscriptSegment: (interviewId: string, data: {
    speaker_role?: string;
    text: string;
    start_time_seconds?: number;
    end_time_seconds?: number;
    confidence?: number;
    is_final?: boolean;
    detected_topics?: string[];
  }) => {
    return apiClient<any>(`/interviews/${interviewId}/adaptive/transcripts`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  getTranscripts: (interviewId: string) => {
    return apiClient<any[]>(`/interviews/${interviewId}/adaptive/transcripts`, {
      method: 'GET',
    });
  },

  generateAdaptiveRecommendation: (interviewId: string, data: {
    session_id?: string;
    current_question?: string;
    candidate_response?: string;
    competency_focus?: string;
    difficulty?: string;
  }) => {
    return apiClient<any>(`/interviews/${interviewId}/adaptive/recommendations/generate`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  getAdaptiveRecommendations: (interviewId: string) => {
    return apiClient<any[]>(`/interviews/${interviewId}/adaptive/recommendations`, {
      method: 'GET',
    });
  },

  acceptAdaptiveRecommendation: (interviewId: string, recId: string) => {
    return apiClient<any>(`/interviews/${interviewId}/adaptive/recommendations/${recId}/accept`, {
      method: 'POST',
    });
  },

  editAdaptiveRecommendation: (interviewId: string, recId: string, editedQuestion: string) => {
    return apiClient<any>(`/interviews/${interviewId}/adaptive/recommendations/${recId}/edit`, {
      method: 'POST',
      body: JSON.stringify({ edited_question: editedQuestion }),
    });
  },

  rejectAdaptiveRecommendation: (interviewId: string, recId: string, reason?: string) => {
    return apiClient<any>(`/interviews/${interviewId}/adaptive/recommendations/${recId}/reject`, {
      method: 'POST',
      body: JSON.stringify({ reason }),
    });
  },

  skipAdaptiveRecommendation: (interviewId: string, recId: string) => {
    return apiClient<any>(`/interviews/${interviewId}/adaptive/recommendations/${recId}/skip`, {
      method: 'POST',
    });
  },

  getLiveCoverage: (interviewId: string) => {
    return apiClient<any>(`/interviews/${interviewId}/adaptive/coverage`, {
      method: 'GET',
    });
  },
};

